src/control/client/config.go

```diff
@@ -0,0 +1,207 @@
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package client
+
+import (
+	"github.com/daos-stack/daos/src/control/common"
+	"github.com/daos-stack/daos/src/control/log"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+)
+
+const (
+	providerEnvKey       = "CRT_PHY_ADDR_STR"
+	ofiInterfaceEnv      = "OFI_INTERFACE"
+	ofiPortEnv           = "OFI_PORT"
+	daosAgentDrpcSockEnv = "DAOS_AGENT_DRPC_SOCK"
+)
+
+// External interface provides methods to support various os operations.
+type External interface {
+	getenv(string) string
+}
+
+type ext struct{}
+
+// runCommand executes command in subshell (to allow redirection) and returns
+// error result.
+func (e *ext) runCommand(cmd string) error {
+	return exec.Command("sh", "-c", cmd).Run()
+}
+
+// getEnv wraps around os.GetEnv and implements External.getEnv().
+func (e *ext) getenv(key string) string {
+	return os.Getenv(key)
+}
+
+// Configuration contains all known configuration variables available to the client
+type Configuration struct {
+	SystemName      string   `yaml:"name"`
+	SocketDir       string   `yaml:"socket_dir"`
+	AccessPoints    []string `yaml:"access_points"`
+	Port            int      `yaml:"port"`
+	CaCert          string   `yaml:"ca_cert"`
+	Cert            string   `yaml:"cert"`
+	Key             string   `yaml:"key"`
+	Provider        string   `yaml:"provider"`
+	FabricIface     string   `yaml:"fabric_iface"`
+	FabricIfacePort int      `yaml:"fabric_iface_port"`
+	LogFile         string   `yaml:"log_file"`
+	Path            string
+	ext             External
+}
+
+// newDefaultConfiguration creates a new instance of configuration struct
+// populated with defaults.
+func newDefaultConfiguration(ext External) Configuration {
+	return Configuration{
+		SystemName:      "daos",
+		SocketDir:       "/var/run/daos_agent",
+		AccessPoints:    []string{"localhost"},
+		Port:            10000,
+		CaCert:          "./.daos/ca.crt",
+		Cert:            "./.daos/daos.crt",
+		Key:             "./.daos/daos.key",
+		Provider:        "ofi+socket",
+		FabricIface:     "qib0",
+		FabricIfacePort: 20000,
+		LogFile:         "/tmp/daos_agent.log",
+		Path:            "etc/daos.yml",
+		ext:             ext,
+	}
+}
+
+// ProcessConfigFile loads a configuration file from the path given,
+// or from the default location if none is provided.  It returns a populated
+// Configuration struct based upon the default values and any config file overrides.
+func ProcessConfigFile(ConfigPath string) (Configuration, error) {
```

Comment:
Rather than passing in ConfigPath, could use opts.ConfigPath directly as opts is global. probably best if we can stop opts being global

Reply:
After seeing your comment, I tried using opts.ConfigPath but was not able to see it as a global within client/config.go.  But based on your second comment, I think it makes sense if we keep the function interface as it is, so that we still have to pass in a value rather than rely on a global.  Maybe I am misinterpreting what you are asking for though.  I'll keep it as-is until we talk again.